docs/proposals/config_data.md

```diff
@@ -0,0 +1,315 @@
+is:
+
+1. A `ConfigData` object is created
+2. A pod that consumes the configuration data via environment variables is created
+3. The pod is scheduled onto a node
+4. The kubelet retrieves the `ConfigData` resource(s) referenced by the pod and starts the container
+   processes with the appropriate data in environment variables
+
+### Consuming `ConfigData` in Volumes
+
+Many programs read their configuration from configuration files.  `ConfigData` should be possible
+to consume in a volume.  The rough series of events for consuming `ConfigData` this way
+is:
+
+1. A `ConfigData` object is created
+2. A new pod using the `ConfigData` via the volume plugin is created
+3. The pod is scheduled onto a node
+4. The Kubelet creates an instance of the volume plugin and calls its `Setup()` method
+5. The volume plugin retrieves the `ConfigData` resource(s) referenced by the pod and projects
+   the appropriate data into the volume
+
+### Consuming `ConfigData`  Updates
+
+Any long-running system has configuration that is mutated over time.  Changes made to configuration
+data must be made visible to pods consuming data in volumes so that they can respond to those
+changes.
+
+The `resourceVersion` of the `ConfigData` object will be updated by the API server every time the
+object is modified.  After an update, modifications will be made visible to the consumer container:
+
+1. A `ConfigData` object is created
+2. A new pod using the `ConfigData` via the volume plugin is created
+3. The pod is scheduled onto a node
+4. During the sync loop, the Kubelet creates an instance of the volume plugin and calls its
+   `Setup()` method
+5. The volume plugin retrieves the `ConfigData` resource(s) referenced by the pod and projects
+   the appropriate data into the volume
+6. The `ConfigData` referenced by the pod is updated
+7. During the next iteration of the `syncLoop`, the Kubelet creates an instance of the volume plugin
+   and calls its `Setup()` method
+8. The volume plugin projects the updated data into the volume
+
+It is the consuming pod's responsibility to make use of the updated data once it is made visible.
+
+Because environment variables cannot be updated without restarting a container, configuration data
+consumed in environment variables will not be updated.
+
+### Advantages
+
+* Easy to consume in pods; consumer-agnostic
+* Configuration data is persistent and versioned
+* Consumers of configuration data in volumes can respond to changes in the data
+
+## Proposed Design
+
+### API Resource
+
+The `ConfigData` resource will be added to the `extensions` API Group:
+
+```go
+package api
+
+// ConfigData holds configuration data for pods to consume.
+type ConfigData struct {
+	TypeMeta   `json:",inline"`
+	ObjectMeta `json:"metadata,omitempty"`
+
+  // Data contains the configuration data.  Each key must be a valid DNS_SUBDOMAIN or leading
+  // dot followed by valid DNS_SUBDOMAIN.
+	Data map[string]string `json:"data,omitempty"`
```

Comment:
Having the keys allowed differ between secret and config data is really
going to confuse users and complicate using them.

I think it's a really important use case to be able to move the same inputs
to secrets and vice versa without problem.  We've talked about how the
base64 encoding sucks for usability on secrets - maybe we should solve that
and then have config also support that.

I would really want to tie config data and secrets more closely, not less.


On Nov 19, 2015, at 12:46 AM, Tim Hockin <notifications@github.com> wrote:

In docs/proposals/config_data.md
<https://github.com/kubernetes/kubernetes/pull/6477#discussion_r45302204>:

> +
> +### API Resource
> +
> +The `ConfigData` resource will be added to the `extensions` API Group:
> +
> +```go
> +package api
> +
> +// ConfigData holds configuration data for pods to consume.
> +type ConfigData struct {
> +	TypeMeta   `json:",inline"`
> +	ObjectMeta `json:"metadata,omitempty"`
> +
> +  // Data contains the configuration data.  Each key must be a valid DNS_SUBDOMAIN or leading
> +  // dot followed by valid DNS_SUBDOMAIN.
> +	Data map[string]string `json:"data,omitempty"`

only change I want to see, then, is that we drop "dot followed by valid
DNS_SUBDOMAIN" as valid keys. Not useful, I think.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/6477/files#r45302204>.